### PR TITLE
Implementation to support Server-side operations in KUIB via Datasource

### PR DESCRIPTION
**Pull request by audaciousanil**
Implementation to support Server-side operations in KUIB via Datasource. Changes include
Datasource
Typings and Transpiled version of Datasource

## PR Checklist

- [ ] See our contributing guidelines: https://github.com/progress/JSDO/blob/master/contributing.md#contribute-to-the-code-base.
- [ ] You have signed the [CLA](https://www.progress.com/jsdo/cla).

[Describe the impact of the changes here. Please include issue # if applicable.]

### CODE REVIEW

**Comment by audaciousanil** [comment]
Thank you Edsel.

Handled result properly in read() API in Datasource

**Comment by audaciousanil** [comment]
Also, added a new **_interface_** named DataResult in the typings file which can be used to handle **data**.

File: packages/ng-datasource/src/progress.data.ng.ds.js [hunk 1]
```diff
@@ -66,10 +66,27 @@ var DataSource = /** @class */ (function () {
...
         }
         this._tableRef = this._options.tableRef;
+        // Find out the name of 'Count' function from Catalog if defined as 'Count' operation
+        // instead of an INVOKE 
+        if (this._options.countFnName !== undefined) {
+            if (typeof (this.jsdo[this._options.countFnName]) !== "function") {
+                throw new Error("Invoke operation '" +
+                    this._options.countFnName + "' for countFnName is not defined.");
+            }
+        }
```
Comment (edselg) [inline packages/ng-datasource/src/progress.data.ng.ds.js]: Use } else {

File: packages/ng-datasource/src/progress.data.ng.ds.js [hunk 2]
```diff
@@ -110,7 +127,26 @@ var DataSource = /** @class */ (function () {
...
                 .then(function (result) {
                 _this._initFromServer = true;
-                resolve(_this.getJsdoData());
+                var data = _this.getJsdoData();
+                if ((_this._options.countFnName && _this._options.countFnName !== undefined) && !(params.skip == 0 && params.top > data.length)) { // Server-side operations
+                    _this.getRecCount(_this._options.countFnName, params)
+                        .then(function (result) {
+                        if (result == undefined && result == null) {
+                            reject(new Error(_this.normalizeError(result, "Unexpected response from 'Count Function' Operation", "")));
+                        }
```
Comment (edselg) [inline packages/ng-datasource/src/progress.data.ng.ds.js]: Use } else {

File: packages/ng-datasource/src/progress.data.ng.ds.ts [hunk 3]
```diff
@@ -75,15 +81,33 @@ export class DataSource {
...
+
+        // Find out the name of 'Count' function from Catalog if defined as 'Count' operation
+        // instead of an INVOKE 
+        if (this._options.countFnName !== undefined) {
+            if (typeof (this.jsdo[this._options.countFnName]) !== "function") {
+                throw new Error("Invoke operation '" +
+                    this._options.countFnName + "' for countFnName is not defined.");
+            }
+        }
+        else if (this.jsdo['_resource'].generic.count !== undefined) {
```
Comment (edselg) [inline packages/ng-datasource/src/progress.data.ng.ds.ts]: Use } else if

File: packages/ng-datasource/typings/progress.data.ng.ds.d.ts [hunk 4]
```diff
@@ -12,21 +12,29 @@ export declare class DataSourceOptions {
     mergeMode?: any;
     pageSize?: any;
     readLocal?: boolean;
+    countFnName?: string;
+}
+export interface DataResult {
```
Comment (edselg) [inline packages/ng-datasource/typings/progress.data.ng.ds.d.ts]: Confirm that tslint is running using tslint.json for NativeScript. If error with "interface" change to "class".

Comment (audaciousanil) [reply packages/ng-datasource/typings/progress.data.ng.ds.d.ts]: I have added tslint.json in my project and it does not complain about the interface

File: packages/ng-datasource/src/progress.data.ng.ds.js (hunk 2 above)
Comment (audaciousanil) [reply packages/ng-datasource/src/progress.data.ng.ds.js]: Handled this

File: packages/ng-datasource/src/progress.data.ng.ds.js (hunk 1 above)
Comment (audaciousanil) [reply packages/ng-datasource/src/progress.data.ng.ds.js]: Handled this

**Pull request merged**